app/src/org/commcare/tasks/SaveToDiskTask.java

```diff
@@ -289,25 +290,21 @@ public void setFormSavedListener(FormSavedListener fsl) {
      * though, until all answers conform to their constraints/requirements.
      */
     private boolean hasInvalidAnswers(boolean markCompleted) {
-        FormIndex i = FormEntryActivity.mFormController.getFormIndex();
-        FormEntryActivity.mFormController.jumpToIndex(FormIndex.createBeginningOfFormIndex());
-
+        FormController formController = FormEntryActivity.mFormController;
```

Comment:
Is there a reason these changes are needed ?

Reply:
Yeah, so the earlier code used to modify the `currentFormIndex` to loop over all the questions. But if at the same time, some other thread also wants to get hold of the current question in screen, it'll have to use the `currentFormIndex` which might give wrong value, if this thread has already modified it. And hence may produce NPE in worst cases. 

Since, we've to loop over all the questions here, we don't really need to modify the `currentFormIndex` rather we can simply have another form index to get all the questions. And this will also make sure that the other threads who might be using the `currentFormIndex` at the same time don't get wrong values.

Reply:
Got it, thanks for the explanation. Only suggestion here would be to abstract this out into it's own PR so that it's easy to revert.